### PR TITLE
Do not form 0*inf evidence in associative memory

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
## Summary\n- Skip exact zero-weight row contributions before multiplication, preventing inactive stored infinities from creating 0 * inf NaNs.\n- Reject non-finite or out-of-domain adaptive controller scalars, including non-boolean adaptive flags and non-integral effective budgets.\n- Treat an update as a fail-closed transaction: validate the complete floating source, prediction, and candidate trees; retain the source state and neutralize outputs when rejected.\n- Expose update_applied on one-step results and updates_applied through the scan runner.\n- Keep corrupted active family logits non-finite and fail-visible; they are no longer disguised as a uniform valid gate.\n\n## Validation\n- [x] 28 focused associative-memory tests\n- [x] 53 tests across associative memory, Step 2 theory invariants, and pipeline integration\n- [x] Ruff on the changed source and tests\n- [x] Repository-wide strict mypy: 163 source files\n\nOriginal contribution made with Cursor. Repair and adversarial validation completed by a maintainer.

### CODE REVIEW

**Review by renzo4web — APPROVED** [review]
Independent verification of the PR #196 scenario — RED confirmed on pristine `main`, GREEN confirmed on this branch, finite path byte-identical, downstream consumers green.

## Reproduction

| Check | pristine `main` (`58b6780`) | this branch (`744c33f`) |
|---|---|---|
| `predict` with an inf stored value in a zero-weight row | evidence sum becomes `0*inf = NaN` → non-finite logits/probabilities | finite logits/probabilities (zero-weight rows skipped before the product) |
| `update` on a poisoned state (inf `values`) | non-finite state committed, non-finite metrics | transaction rejected: source state retained, predictions/logits/metrics neutralized to zeros, `update_applied=False` |
| Adaptive-family `predict` with inf `family_logits` | NaN gate output | stays non-finite and fail-visible (no uniform-simplex disguise) |
| Config `write_lr=inf` (and 7 more adaptive scalars) | accepted | rejected at construction with `ValueError` |
| `pytest tests/test_associative_memory.py` | (bug unfixed) | **28 passed** |
| `pytest tests/test_step2_theory_invariants.py tests/test_pipeline.py` (signature-change consumers) | — | **25 passed** (integration, branch overlay) |
| `ruff check .` | — | clean |
| `mypy` (repo config) | — | clean, 163 files |

RED probe run against a pristine `main` worktree with the repo `.venv` (jax 0.11.0); GREEN run in an isolated sandbox (Docker `python:3.12-slim`, `--network none`, read-only root, `--cpus 2 --memory 2g --pids-limit 128`, `pip install -e ".[dev]"`), overlaying exactly this branch's `core/associative_memory.py` + `tests/test_associative_memory.py` onto the `58b6780` tree.

## Finite-path bit-exactness (the #43 lesson, checked)

Deterministic probe: a 100-step predict+update chain (finite contexts/labels), hashing every output leaf (logits, probabilities, weights, scope/base weights, family/window probs, budget, state values/prior/utility, metrics):

```text
main   finite-path digest: d01b70689536d43c0db705d6fc403dc16f041a2863a008985653ca91872ecc03
branch finite-path digest: d01b70689536d43c0db705d6fc403dc16f041a2863a008985653ca91872ecc03
```

Byte-identical across two runs per side; cross-module probes also match `main` digests under this overlay, so the change is inert outside `associative_memory`.

Design notes (not blockers):
- The evidence-sum guard only skips exact zero weights — a tiny-but-nonzero weight times inf still yields the fail-visible inf/NaN path, so no silent clamping was introduced.
- `update` is now a fail-closed transaction via the shared `update_safety.py` helpers (`floating_tree_is_finite`, `select_transaction`, `neutralize_array`), and the transaction verdict is exported (`update_applied`, `updates_applied` through `run_associative_memory_arrays`). This is an API addition; existing consumers (`steps/step2.py`, `pipeline.py`) construct `AssociativeMemoryUpdateResult` positionally and remain green under the branch overlay.
- The new config gate widens validation (non-finite, non-boolean, non-integral scalars rejected at construction). Defaults are untouched, so pinned runs and evidence lanes are unaffected.

Probe: <details><summary>finite-path digest probe (PR196 section)</summary>

```python
import hashlib
import jax.numpy as jnp
from alberta_framework.core.associative_memory import (
    AssociativeMemoryConfig,
    AssociativeMemoryLearner,
)

m = AssociativeMemoryLearner(
    AssociativeMemoryConfig(vocab_size=4, block_size=3, suffix_length=2, max_features=8)
)
state = m.init()
h = hashlib.sha256()
for i in range(100):
    ctx = jnp.asarray([(i % 4), ((i + 1) % 4), ((i + 2) % 4)], dtype=jnp.int32)
    label = jnp.asarray(i % 4, dtype=jnp.int32)
    pred = m.predict(state, ctx)
    for leaf in (pred.logits, pred.probabilities, pred.weights, pred.scope_weights,
                 pred.base_weights, pred.family_probs, pred.window_probs,
                 pred.total_weight, pred.effective_budget):
        h.update(jnp.asarray(leaf, dtype=jnp.float32).tobytes())
    res = m.update(state, ctx, label)
    for leaf in (res.logits, res.predictions, res.metrics, res.state.values,
                 res.state.prior, res.state.utility):
        h.update(jnp.asarray(leaf, dtype=jnp.float32).tobytes())
    state = res.state
print("finite-path digest:", h.hexdigest())
```

</details>

AI-assisted review (provider: zai/glm-5.3 via opencode-go; no Slop receipt applies to review work).

slop-review {"schemaVersion":"1","projectId":"asi","artifactUrl":"https://github.com/elizaOS/asi/pull/196","headSha":"744c33fc255e86a5cf7d2c012681aa1d4e9033ca","recommendation":"accept","reproduced":true,"securityRisk":"none","duplicateRisk":"none","usefulArtifacts":[],"commands":["git worktree add /tmp/revA-main 58b6780",".venv/bin/python /tmp/revA-meta/red_probes.py","docker run --rm --network none --read-only --tmpfs /tmp:rw,size=512m --cpus 2 --memory 2g --pids-limit 128 reva-verify:pr196 bash /verify_pr.sh 196 core/associative_memory.py tests/test_associative_memory.py","docker run --rm --network none --read-only --tmpfs /tmp:rw,size=512m --cpus 2 --memory 2g --pids-limit 128 reva-verify:pr196 python -m pytest tests/test_step2_theory_invariants.py tests/test_pipeline.py -q -o addopts= -p no:cacheprovider","python /probe_all.py PR196"],"evidenceUrls":["https://github.com/elizaOS/asi/pull/196"],"summary":"RED reproduced on pristine main 58b6780 (0*inf=NaN evidence from zero-weight inf rows, non-finite state committed by update, write_lr=inf accepted); GREEN in a no-network Docker sandbox: 28/28 module tests plus step2/pipeline integration tests pass, ruff and mypy clean, and a 100-step deterministic predict+update finite-path digest is byte-identical between main and branch (d01b7068...). Transaction fail-closed semantics use the shared update_safety helpers; no evidence-manifest, pinned-output, threshold, or validator files touched."}
